### PR TITLE
Switch from Parcel to Webpack

### PR DESCRIPTION
**Pull request by JakeMakesStuff**
## Type of Change

Adding a Parcel-like dev env and switching the build code to Webpack.

## What issue does this relate to?

N/A

### What should this PR do?

Switch to Webpack.

### What are the acceptance criteria?

Does this behave as expected? [Latest commit on DNS tool.](https://github.com/do-community/dns-tool/commit/89a31ca087007c5572738e3f988311d37dbff730#commitcomment-46860916)

### CODE REVIEW

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
👋 I've been testing this out with the bandwidth-tool -- my big issue with this change currently is that in dev, we appear to not be using HMR.

This means that making a single copy edit anywhere in the tool requires the whole tool to be rebuilt, resulting in needing to wait for ~50s for it to rebuild before the single copy change can be seen.

With Parcel in master, the same copy edit takes ~1.2s to build and be rendered.

File: demo/src/in/a/path/far/far/away/increment.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// Defines the current count.
```
Comment (MattIPv4) [inline demo/src/in/a/path/far/far/away/increment.js:1]: License header pls, for consistency :)

File: demo/src/scss/style.scss [hunk 2]
```diff
@@ -0,0 +1 @@
+/* Content would go here if this was a real application. */
```
Comment (MattIPv4) [inline demo/src/scss/style.scss:1]: Ditto re: adding license header

File: package.json [hunk 3]
```diff
@@ -4,7 +4,12 @@
   "description": "Custom Vue templates & resources for DigitalOcean Community tools.",
   "main": "src/cli.js",
   "scripts": {
-    "test": "eslint './src/**/*.{vue,js}'"
+    "test": "eslint './src/**/*.{vue,js}' && npm run demo:build",
```
Comment (MattIPv4) [inline package.json:7]:
```suggestion
    "test": "npm run test:eslint && npm run demo:build",
    "test:eslint": "eslint './src/**/*.{vue,js}'",
    "test:eslint:fix": "npm run test:eslint -- --fix",
```
Consistent with how we do these scripts elsewhere

File: package.json [hunk 4]
```diff
@@ -4,7 +4,12 @@
...
+    "demo:init": "cd demo && node ../src/cli.js template",
+    "demo:dev": "npm run demo:init && cd demo && node ../src/cli.js dev ./src ./dev 8000",
+    "demo:build": "npm run demo:init && cd demo && node ../src/cli.js tool ./src ./dist",
+    "deploy:spaces:comment": "node ./src/cli.js comment do-vue",
+    "lint-and-fix": "eslint './src/**/*.{vue,js}' --fix"
```
Comment (MattIPv4) [inline package.json:12]:
```suggestion
```
Moved up

File: src/build/buildTool.js [hunk 5]
```diff
@@ -14,106 +14,34 @@ See the License for the specific language governing permissions and
 limitations under the License.
```
Comment (MattIPv4) [inline src/build/buildTool.js:14]: License header year needs updating

File: src/build/createGitHubComment.js [hunk 6]
```diff
@@ -18,7 +18,10 @@ const fetch = require('node-fetch');
```
Comment (MattIPv4) [inline src/build/createGitHubComment.js:18]: Ditto re: updating license header

File: src/build/dev.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+*/
+
+const config = require('../webpack.config');
+const webpack = require('webpack');
+const path = require('path');
+const WebpackDevServer = require('webpack-dev-server');
+
+module.exports = (source, out, port) => {
+    const abs = x => path.join(process.cwd(), path.normalize(x));
+    out = abs(out);
```
Comment (MattIPv4) [inline src/build/dev.js:24]:
```suggestion
    const contentOut = abs(out);
```
I'd generally try to avoid mutating args

File: src/build/dev.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    const compiler = webpack(config(abs(source), out, true));
```
Comment (MattIPv4) [inline src/build/dev.js:25]:
```suggestion
    const compiler = webpack(config(abs(source), contentOut, true));
```
Based on above suggestion

File: src/build/dev.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    const server = new WebpackDevServer(compiler, {
+        contentBase: out,
```
Comment (MattIPv4) [inline src/build/dev.js:27]:
```suggestion
        contentBase: contentOut,
```
Based on above suggestion

File: src/build/index.js [hunk 10]
```diff
@@ -14,11 +14,12 @@ See the License for the specific language governing permissions and
...
-const createGitHubComment = require('./createGitHubComment');
-const ensureDir = require('./ensureDir');
-const fetchTemplate = require('./fetchTemplate');
+const buildSVGs = source => require('./buildSVGs')(source);
+const buildTool = (source, out) => require('./buildTool')(source, out);
+const cleanDist = () => require('./cleanDist')();
+const createGitHubComment = args => require('./createGitHubComment')(args);
+const ensureDir = dir => require('./ensureDir')(dir);
+const fetchTemplate = () => require('./fetchTemplate')();
+const dev = (source, out, port) => require('./dev')(source, out, Number(port));
```
Comment (MattIPv4) [inline src/build/index.js:23]: Why? We seem to be wrapping this is anon functions that are just passing the args through, which seems rather redundant?

Only diff being the `Number` wrap on port, which we can do directly in `cli.js`?

File: src/build/index.js [hunk 11]
```diff
@@ -14,11 +14,12 @@ See the License for the specific language governing permissions and
 limitations under the License.
```
Comment (MattIPv4) [inline src/build/index.js:14]: Ditto re: updating license header

File: src/webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
+const webpack = require('webpack');
```
Comment (MattIPv4) [inline src/webpack.config.js:1]: Ditto re: adding license header

File: src/webpack.config.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        },
+        extensions: ['.js', '.jsx', '.json', '.ts', '.tsx', '.css', '.scss', '.sass', '.html', '.vue', '.yml', '.yaml'],
+    },
+    plugins: [
+        new webpack.ProvidePlugin({
+            process: 'process/browser.js',
+            Buffer: ['buffer', 'Buffer'],
+        }),
+        new webpack.HotModuleReplacementPlugin(),
+        new BundleAnalyzerPlugin({analyzerMode: 'static', openAnalyzer: false}),
```
Comment (MattIPv4) [inline src/webpack.config.js:114]:
```suggestion
        new BundleAnalyzerPlugin({ analyzerMode: 'static', openAnalyzer: false }),
```
Object spacing

File: src/webpack.config.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        new VueLoaderPlugin(),
+        new MiniCssExtractPlugin({filename: 'style.css'}),
```
Comment (MattIPv4) [inline src/webpack.config.js:116]:
```suggestion
        new MiniCssExtractPlugin({ filename: 'style.css' }),
```
Object spacing

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
Assorted code cleanup things -- no actual functionality changes that should need the tools to be bumped

File: src/build/index.js (hunk 10 above)
Comment (JakeMakesStuff) [reply src/build/index.js:23]: This is to make debugging easier. If we import them all at once, a lot of require noise gets pushed in for every command, meaning one command that would normally perfectly run can error due to another file.

Comment (MattIPv4) [reply src/build/index.js:23]: I wonder if we can actually just remove this file completely? It seems that everything here is also a 1:1 map with file names, which we can just directly import where needed?

**Review by MattIPv4 — COMMENTED** [review]
Preliminary lgtm pending https://github.com/do-community/minify-tool/pull/9#pullrequestreview-609963833 being resolved (or not if we deem it not worth blocking this)

**Review by MattIPv4 — APPROVED** [review]
Minify-tool deps issue is a no-fix for now, so this is good to go! :shipit: 

Let's get this merged and then start unpinning all the tool migration branches :)

**Pull request merged**